README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![](https://img.shields.io/github/issues/wangjiaxuan666/autopca)](https://img.shields.io/github/issues/wangjiaxuan666/autopca)
[![](https://img.shields.io/github/issues/wangjiaxuan666/autopca)](https://img.shields.io/github/forks/wangjiaxuan666/autopca)
[![](https://img.shields.io/github/issues/wangjiaxuan666/autopca)](https://img.shields.io/github/stars/wangjiaxuan666/autopca)
<!-- badges: end -->

# autopca:wrench: <img src='man/figures/autopca_logo.png' align="right" height="139" />

**Author**: [Jiaxuan Wang](https://abego.cn/) :pencil:

[![](https://img.shields.io/badge/follow%20me%20on-WeChat-green.svg)](https://pic-1259340288.cos.ap-guangzhou.myqcloud.com/back/qrcode_for_Wechat.jpg)

Originally,this R package `autopca` was a script I used to draw PCA. PCA
can analyze batch effects in experimental processing, and also analyze
experimental processing factors. In my point, it is very important in
checking omics data. At present, this R package Features are very few
But are under active development. The basic function is already there,
so I placed on `github`, if necessary, you can download and use it
yourself.

## Update

``` r
install.packages("devtools")
devtools::install_github("wangjiaxuan666/autopca")
```

## Installation

The R packages only can install the development version from
[GitHub](https://github.com/) with:

Notice : if you install `autopca` failed , please check the R installed
envirment. maybe you need install `tidyverse` or `plyr` before. IF YOU
FAILED WITH ANGRY AND BOOM ERRORS. Please contact me with
<poormouse@126.com> , I’ m sorry about what happen to you. Although it’s
me, I also have to admit that this is a bug-filled R package.

> Every time I use this R packages , it will spend 95% time to fix new
> bug! Although it will spend a little time if i use the scicpt not R
> packages. Automation is always very difficult because it is suitable
> for all kinds of situations and data type. BUT IT IS A FUN WHEN A
> PROBLEM NEED TO THINK

## Example

Before we start the PCA analysis, we first need to tidy the input data.
We can use the function `pca_data_tidy`to make the data clean for the
next PCA analysis.It should be noted that the format of the input data
is very important to determine the success of subsequent analysis.

Here, Emphasize the format of the input data. First, PCA analysis use
the function `stat::prcomp`, Consequently the result given the every
observed values’ Variance in principal components. the observed values
is the `prcomp` data’ rownames. For example, the data iris

``` r
head(iris)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
#> 6          5.4         3.9          1.7         0.4  setosa
```

the rownames is every iris ID number, in iris, we want to demonstrate
the proportion of each iris flower in terms of petal length and width
and so on. That is we want, so for iris data, we just tidy a little for
next analysis. Just like

``` r
library(autopca)
library(patchwork)
irisgroup <- iris[,5] 
iristidy <- iris[,-5]
p1 = pca(iris[,-5],sample_group = as.data.frame(iris[,5]))
p2 = pca(iris[,-5],sample_group = as.data.frame(iris[,5]),add_ploy =  "ellipse")
p3 =pca(iris[,-5],sample_group = as.data.frame(iris[,5]),add_ploy = "encircle")
p4 = pca(iris[,-5],sample_group = as.data.frame(iris[,5]),add_ploy = "polygon")
p1+p2+p3+p4
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" /> the
group information is Grouping information is required, otherwise an
error will be reported!

``` r
#pca(iris)
```

As for why and how to use `autopca`, We need to start from the
beginning.

## Illustration

the `autopca` designed for transcriptome data, The classic transcriptome
data is rownames is gene id, the column names is every observed sample.
sometime it also add some annotation information in the tail. like this.

``` r
test = matrix(rnorm(200), 20, 10)
test[1:10, seq(1, 10, 2)] = test[1:10, seq(1, 10, 2)] + 3
test[11:20, seq(2, 10, 2)] = test[11:20, seq(2, 10, 2)] + 2
test[15:20, seq(2, 10, 2)] = test[15:20, seq(2, 10, 2)] + 4
colnames(test) = paste("Test", 1:10, sep = "")
rownames(test) = paste("Gene", 1:20, sep = "")
annot <- c(rep("KEGG",20))
test <- data.frame(test,annot)
head(test)
#>          Test1       Test2    Test3       Test4    Test5
#> Gene1 3.885105 -1.40996356 2.082833  1.74326260 2.922234
#> Gene2 3.116096 -0.79413555 3.188819 -0.40254010 3.036880
#> Gene3 2.817635 -0.98390401 2.651808 -1.87837298 2.957372
#> Gene4 2.696947 -0.05363688 3.058046  0.13078877 3.612569
#> Gene5 1.544220  0.34743798 2.642220 -0.04920413 1.315197
#> Gene6 4.434048 -0.42054450 3.187126 -2.05952773 2.736322
#>             Test6    Test7       Test8    Test9       Test10
#> Gene1  1.08305670 4.035495 -0.01374523 5.178203  0.001947392
#> Gene2  0.23804056 2.429948  1.40977634 4.485687  0.571390923
#> Gene3 -0.32113599 2.343335  0.01128968 1.884459  0.230932915
#> Gene4 -0.06159268 3.369248  0.13621463 4.633204 -0.476862998
#> Gene5 -1.08884753 3.528223 -1.01294805 2.711387  0.075084606
#> Gene6  0.09278681 2.462118 -0.40843086 2.840577 -0.116544660
#>       annot
#> Gene1  KEGG
#> Gene2  KEGG
#> Gene3  KEGG
#> Gene4  KEGG
#> Gene5  KEGG
#> Gene6  KEGG
```

Through the above steps, we obtained a classic transcriptome data frame.
**NOW WE explain the sample variance in PC. So we need tidy the data by
the function`pca_data_tidy`.**

``` r
pca_data_tidy(as.data.frame(test)) -> test_tidy
#> ... Notice: the input data is a data frame not a tibble
#> ... Problem: the input data vaule in every column must be numberic value
#> ... Problem: the error because the character type value in data
#> ... Successed! the pca data save in the object
```

NEXT, we can use the tidy data to analysis, just like this

``` r
as.data.frame(c(rep("A",5),rep("B",5))) -> group
rownames(group) <- colnames(test[,-11])
colnames(group) <- "group"
head(group)
#>       group
#> Test1     A
#> Test2     A
#> Test3     A
#> Test4     A
#> Test5     A
#> Test6     B
pca(test_tidy,sample_group = group)
```

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" />

But `pca` function not only that, It supports regular matching
characters to replace the names of sample or group. When the sample
names is “CK-1\_fokm, CK-2\_fokm,” and so on , It will be very useful.

``` r
#pca(data = re,# the data didn't exist,just a example to display the parameter
#    display_sample = TRUE,
#    rename = "replace",
#    str_sample = "_.*",
#    str_group = "-\\d$",
#    add_ploy = TRUE)
```

ALL parameters explain :

**Usage**:

    pca(
      data = data,
      center = T,
      retx = T,
      scale = FALSE,
      display_sample = FALSE,
      rename = c("diy", "replace"),
      sample_group = NULL,
      str_sample = NULL,
      str_group = "-.*",
      add_ploy = FALSE
    )

**Arguments**

    data:iput data form the function 'pca_data_tidy'

    center:the prcomp param, detail see '?prcomp'

    retx:the prcomp param, detail see '?prcomp'

    scale:the prcomp param, detail see '?prcomp'

    display_sample:if TRUE will add the text labels on points.

    rename：the method for change the sample and group names, two argment can choose, "diy" is for the creat a data for name,"replace" is use regexp to replace or change the name

    sample_group：a data for change the sample and group name, the rownames is sample and the first column is group

    str_sample:the 'regexp' for the sample name to become the target name

    str_group:the 'regexp' for the group name to become the target name

    add_ploy:if TRUE will add the polygon on points.

Of course, as a fan of `tidyverse`, all function in `autopca` also
support `tibble` data input. If there are any questions and suggestions
in use, welcome questions and suggestions

2021-1-2


NEWS.md
# autopca 0.0.900

* pca function parameter `add_ploy` change to one character of 
  ellipse","encircle","polygon". the plot see the example.

* use `pkgdown` creat a website for introduce this R packages.


test.R
r"(C:\Users\woney\Desktop\一个有意思的项目)"-> pathway
setwd(pathway)
dat = readr::read_tsv("all.genes.expression.annot(2).xls")
require(tidyverse)
dat %>% dplyr::select(ends_with("fpkm")) %>% pca_data_tidy() -> re
pca(data = re,
    display_sample = TRUE,
    rename = "replace",
    str_sample = "_.*",
    str_group = "-\\d$",
    add_ploy = TRUE)



tests/testthat.R
library(testthat)
library(autopca)


R/pca_data_tidy.R
#' the Principal Components Analysis need the data must be transformed
#'
#' @param data the input data must be a dataframe or tibble which' rownames is gene id and colnames is sample name
#' @param id the rowid, can be a character or the column number
#'
#' @return a dataframe
#' @export
#'
#' @examples
#' test = matrix(rnorm(200), 20, 10)
#' test[1:10, seq(1, 10, 2)] = test[1:10, seq(1, 10, 2)] + 3
#' test[11:20, seq(2, 10, 2)] = test[11:20, seq(2, 10, 2)] + 2
#' test[15:20, seq(2, 10, 2)] = test[15:20, seq(2, 10, 2)] + 4
#' colnames(test) = paste("Test", 1:10, sep = "")
#' rownames(test) = paste("Gene", 1:20, sep = "")
#' annot <- c(rep("KEGG",20))
#' test <- data.frame(test,annot)
#' pca_data_tidy(as.data.frame(test)) -> test
pca_data_tidy <- function(data,id = NULL){
  if(is.data.frame(data)){
    if(tibble::is_tibble(data)){
      data = data
      cat("...","Notice:","the input data is a tibble\n")
    } else {
      cat("...","Notice:","the input data is a data frame not a tibble\n")
      data = tibble::rownames_to_column(data)
      data = tibble::as_tibble(data)
      id = "rowname"
    }

    if(is.null(id)){
      data_new =  data[,-1]
      geneid = data[,1]
    } else {
      if(is.numeric(id)){
        data_new =  data[,-id]
        geneid = data[,id]
      } else {
        data_new = data[,-grep(id,colnames(data),ignore.case = TRUE)]
        geneid =  data[,grep(id,colnames(data),ignore.case = TRUE)]
      }
    }

    if(!all(unlist(lapply(data_new, is.numeric)))){
      cat("...","Problem:","the input data vaule in every column must be numberic value\n")
      cat("...","Problem:","the error because the character type value in data\n" )
      data_new = data_new[,unlist(lapply(data_new, is.numeric))]
    } else {
      cat("...","Wonderful:","the input data vaule in every column all be numberic value\n")
    }

    data_id <- cbind(geneid,data_new)
    pca_data=t(as.matrix(data_id))
    exp_data <- pca_data[-1,]
    mode(exp_data)="numeric"
    cat("...","Successed!","the pca data save in the object","\n")
    return(exp_data)

  } else {
    stop("...","Failed!!!:","please input the data that must be a data.frame or a tibble")
  }
}


R/pca.R
#' PCA plot
#'
#' @param data iput data form the function 'pca_data_tidy' or a matrix of column(gene) and row(sample).
#' @param center the prcomp param, detail see '?prcomp'
#' @param retx the prcomp param, detail see '?prcomp'
#' @param scale the prcomp param, detail see '?prcomp'
#' @param rename the very important parameter,if don't input the rename parameter, will take it as no group.
#' and rename default is "diy", the method for change the sample and group names, two argment
#' can choose, "diy" is for the creat a data for name use the parameter "sample_group", if rename = "replace" is use regexp to
#' replace for change the name by the parameter "str_sample" and "str_group".
#' @param str_sample need the parameter only when rename = "replace", the 'regexp'
#' for the sample name to become the target name
#' @param str_group need the parameter only when rename = "replace", the 'regexp'
#' for the group name base on the sample names haved changed by "str_sample".
#' @param sample_group need the parameter only when rename = "diy", a data frame
#'  for change the sample and group name, the rownames is sample and the first
#'  column is group. or a tibble ,the first column is sample name and the second column is group names.
#' @param display_name if TRUE will add the text labels on points.
#' @param add_ploy will add the ploy line to encircle the sample points, can
#' choose the ploy style form "ellipse","encircle","polygon", see the plot in
#' example. the default is "polygon".
#' @param name_size the font size of sample name.
#' @param point_size the size of the piont
#'
#' @return a ggplot object
#' @export
#' @importFrom ggplot2 ggplot aes geom_point geom_vline geom_hline theme_bw labs theme element_text element_line theme geom_polygon stat_ellipse
#' @importFrom dplyr mutate_at vars
#' @importFrom tibble is_tibble as_tibble
#' @importFrom stats prcomp
#' @importFrom ggalt geom_encircle
#'
#' @examples library(autopca)
#' pca(iris[,-5],sample_group = as.data.frame(iris[,5]))
#' pca(iris[,-5],sample_group = as.data.frame(iris[,5]),add_ploy = "ellipse")
#' pca(iris[,-5],sample_group = as.data.frame(iris[,5]),add_ploy = "encircle")
#' pca(iris[,-5],sample_group = as.data.frame(iris[,5]),add_ploy = "polygon")

pca <- function(data = data,
                center = T,
                retx = T,
                scale = FALSE,
                display_name = FALSE,
                rename = c("diy","replace"),
                sample_group = NULL,
                str_sample = NULL,
                str_group = "-.*",
                add_ploy = c("null","ellipse","encircle","polygon"),
                name_size = 4,
                point_size = 2
){
  info = stats::prcomp(data,center = center,retx = retx,scale. = scale)
  pca_info = summary(info)
  tmp = pca_info$importance
  pc = pca_info$x
  sample = rownames(data)
  rename = match.arg(rename)
  ploy_style = match.arg(add_ploy)

  if(rename == "replace"){
    if(is.null(str_sample)){
      stop("the 'str_sample' argment must be existed")
    } else {
      sample <- sub(str_sample,"",sample)
      group <- sub(str_group,"",sample)
      if_n = length(unique(sample))/length(unique(group))
    }
  }

  if(rename == "diy"){
    if(is.null(sample_group)){
      sample = rownames(data)
      group = rownames(data)
      cat("...Notice:","the sequence of sample names\n")
      cat("...must be same with the input data rownames\n")
    } else {
      if(is.data.frame(sample_group)){
        if(tibble::is_tibble(data)){
          sample = sample_group[,1]
          group = sample_group[,2]
        } else {
          sample = rownames(sample_group)
          group = as.character(sample_group[,1])
        }
      } else {
        cat("...the sample_group must be a dataframe or tibble\n")
      }
    }
    if_n = length(unique(sample))/length(unique(group))
  }

  rownames(pc) <- sample
  pc <- cbind(pc,group,sample)

  if(all(rownames(pc) == sample)){
    pc <- tibble::as_tibble(pc)
    pc <- dplyr::mutate_at(pc, dplyr::vars("PC1","PC2"),as.double)
  } else {
    stop("ERROR!!!!!!, the sample name sequece is not match the input data rowname")
  }
  plot <- ggplot2::ggplot(data = pc,mapping = ggplot2::aes(x = PC1,y = PC2))+
    ggplot2::geom_point(size = point_size,mapping = ggplot2::aes(color = group))+
    ggplot2::geom_hline(yintercept = 0,linetype = 4,color = "grey")+
    ggplot2::geom_vline(xintercept = 0,linetype = 4,color = "grey")+
    ggplot2::labs(
      x = paste("PC1","(",as.numeric(sprintf("%.3f",tmp[2,1]))*100,"%)",sep=""),
      y = paste("PC2","(",as.numeric(sprintf("%.3f",tmp[2,2]))*100,"%)",sep=""),
      title = "PCA")+
    ggplot2::theme_bw()+
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5,size =12))

  if(display_name == FALSE){
    plot = plot
  } else {
    plot = plot + ggplot2::geom_text(ggplot2::aes(label = sample),size = name_size)
  }

  # add plot line
  if(!ploy_style == "null"){
    # must be first vector and select from three input
    if(ploy_style[1] %in% c("ellipse","encircle","polygon")){
      ploy_style = ploy_style[1]
    } else {
      ploy_style = "polygon"
    }
    # three sample can't caculate ellipse will change to
    if(if_n <= 3 & ploy_style == "ellipse"){
      ploy_style = "polygon"
    }

    if(ploy_style == "ellipse"){
      plot = plot + ggplot2::stat_ellipse(geom = "polygon", mapping = ggplot2::aes(fill = group),type = "t", level = 0.95, linetype = 2,alpha = 0.2)
    }

    if(ploy_style == "encircle"){
      plot = plot + ggalt::geom_encircle(ggplot2::aes(group = group,fill = group),alpha=0.4)
    }

    if(ploy_style == "polygon"){
      plot = plot + ggalt::geom_encircle(ggplot2::aes(group = group,fill = group),alpha=0.4,s_shape=1, expand=0)
    }
  }
  # end
  return(plot)
}
